Add cache behavior upfront, and note that priming is over HTTPS

konklone opened the pull request:
This adds the HSTS cache check into the topmost description, so that newcomers to the idea don't mistakenly get the impression that this will cause browsers to issue endless HTTPS HEAD requests for every observed insecure resource.

It also makes explicit that the HSTS priming request is over HTTPS.

konklone committed:
add cache behavior upfront, and note that priming is over HTTPS

mikewest commented on index.src.html:
```diff
@@ -130,8 +130,16 @@ <h3 id="examples">Example</h3>
   Before blocking this insecure request as mixed content, the user agent
-  performs an HSTS priming request to the root of the resource's host, as
-  follows:
+  consults its <a lt="matches the HSTS cache">HSTS cache</a> for the origin of
```
I'd suggest something like "the user agent checks whether the request <a>matches its HSTS cache</a>", and leave out details about origins or hosts (we match against the latter, not the former).

mikewest commented on index.src.html:
```diff
@@ -130,8 +130,16 @@ <h3 id="examples">Example</h3>
+  consults its <a lt="matches the HSTS cache">HSTS cache</a> for the origin of
+  the resource.
+
+  If the origin of the resource matched the HSTS cache, the user agent can then
```
Nit: The cache is also negative; that is, it stores HSTS priming failures as well as successes. "Matching" just means that we don't need to do a priming request, not that we can safely upgrade the request.

konklone replied:
How is cache lifetime determined for priming failures? User agent discretion?

mikewest replied:
Step 5 of http://mikewest.github.io/hsts-priming/#modifications-fetch-priming. Basically "Do whatever." :)

Maybe ~24 hours is reasonable, maybe we should look at the response's cache headers, maybe maybe maybe. _shrug_ Suggestions welcome. :)

mikewest commented on index.src.html:
```diff
@@ -130,8 +130,16 @@ <h3 id="examples">Example</h3>
+  safely upgrade the initial request from HTTP to HTTPS, rather than blocking it
+  as mixed content.
+
+  If the origin of the resource did not match the HSTS cache, the user agent
```
This isn't accurate; see above.

mikewest commented:
Thanks for the PR! I added some nits inline.

konklone commented:
@mikewest Thanks, those were great points all. I updated the PR to remove mention of the origin, since that wasn't accurate, and simplified the description to reflect the fact that the cache check always happens, and the priming request only happens if it's not in the cache.

Src changes, without whitespace changes: https://github.com/mikewest/hsts-priming/pull/2/files?w=1#diff-117d6498d2aa8019cc0abf5eeb87a9fa

Also, it looks like the `bikeshed` process updated the spec's date from 11/6 to 11/7. I can make it so this PR only affects the `.src.html` file, if you want, or make other changes to my workflow.

mikewest commented on index.html:
```diff
@@ -1128,8 +1128,10 @@ <h3 class="heading settled" data-level="1.1" id="examples"><span class="secno">1
     <p>Before blocking this insecure request as mixed content, the user agent
-  performs an HSTS priming request to the root of the resource’s host, as
-  follows:</p>
+  checks whether the request <a data-link-type="dfn" href="#matches-the-hsts-cache">matches its HSTS
```
Nit: You can write this as `<a lt="matches the HSTS cache">matches its HSTS cache</a>`. `lt` => "linking text". I might even suggest `<a lt="matches the HSTS cache">matches</a> its <a>HSTS cache/a>` if you're feeling particularly informative.

konklone replied:
> Nit: You can write this as `<a lt="matches the HSTS cache">matches its HSTS cache</a>`.

But I did! You're looking at the generated `index.html`. In `index.src.html`, I used the `lt` syntax, mimicking what I saw you had done in your draft:

https://github.com/mikewest/hsts-priming/pull/2/files#diff-117d6498d2aa8019cc0abf5eeb87a9faR133

mikewest commented:
LGTM other than the nit I added. If you address that one way or another, and then squash your commits, I'll happily land it for you. :)

konklone commented:
As mentioned above, I believe the Nit was not a Nit, and I've squashed my 2 commits into 1.

mikewest commented:
Ha, you're right. Sorry about that confusion. Thanks for squishing, merging it in now.
